### ISSUE TITLE
Add @BeforeAll and @AfterAll hooks 

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/cucumber-jvm#515]
People keep asking for `BeforeAll` and `AfterAll` hooks. I haven't had a big need for them myself, but it seems to me this would be sufficient:

``` java
public class GlobalHooks {
    private static boolean dunit = false;

    @Before
    public void beforeAll() {
        if(!dunit) {
            Runtime.getRuntime().addShutdownHook(afterAllThread);
            // do the beforeAll stuff...
            dunit = true;
        }
    }
}
```

If this doesn't cut it for you, please explain in this ticket, and maybe we'll add special support for it.

**Comment by eelessam** [comment]
Is anyone working on this at the moment?

**Comment by fernandomora** [comment]
It this feature in the roadmap in a near future?

I have found the need of this feature several times in different projects.
And neither addShutdownHook nor @AfterClass covered my needs, since we execute tests from sbt, addShutdownHook is only executed on JVM shutdown, while this does not happen until sbt is explicitly closed and I don't want to executed our tests in a forked JVM.

**Comment by beemi** [comment]
Hi all , 
please provide solution for this .
I got the feature file which got Background , and 5     Scenario Outline: , i would like to run my Background just only once then i want to execute all my  Scenario Outline: s with the Example data ,

because of the 
@Before
    public void before(Scenario scenario) {
        new CucumberGlobalHooks().before(scenario);
    }


    @After
    public void after(Scenario scenario) {
        new CucumberGlobalHooks().after(scenario);
    }

 my Background executed for each and every  Scenario Outline at the moment , but is there any way to set Just Background execute only once then execute all my Scenario Outlin 's 

please provide me any help full information on this 
Thank you

**Comment by ovictor** [comment]
+1 for this feature.  Is anyone working on it?

**Comment by markwinspear** [comment]
And another +1 from me. I can't use Junit @Before and @After annotations with Cucumber which prevents me from using Cucumber with WebDriver using Page Object Model and executing tests in Saucelabs

**Comment by ryantenney** [comment]
:+1: here as well

**Comment by vaughandroid** [comment]
+1
No word for over a year though, so I'm not holding my breath...

**Comment by dkowis** [comment]
Pull requests accepted, but they need to cooperate with Dependency Injection.

Additionally it might be a feature that comes in when Gherkin 3 gets integrated. Glad to know people want it, but as with any OpenSource project, it moves at the speed of the volunteers that work on it.

**Comment by AniteSridhar** [comment]
Hi All,

I am having 5 Cucumber JUnit runner classes for parallel test execution, I want to concatenate all the JSON outputs together. Is there any Cucumber annotation that can work after all the parallel executions or can anyone suggest me other options which suits my requirement.

Thanks in advance,
Anite S

**Comment by dkowis** [comment]
@AniteSridhar This is not a support forum. See: https://cucumber.io/support

**Comment by stijnvanbael** [comment]
+1

**Comment by dattfemi** [comment]
Junit BeforeClass and AfterClass annotations work fine if you are using a JUnit runner. No issues with the static state.. Works fine with Page object Model for me.

**Comment by andresd24** [comment]
I don't see a resolution for this.  People who have mentioned to use  @AfterClass don't understand what we're trying to solve here.  We want a method that executes after *the entire cucumber test run*.  The @AfterClass doesn't work for this.  It throws errors.   If anyone has a solution it would be awesome if they take the time to share.  thanks.  -AD

**Comment by OlufemiAdeOlusile** [comment]
@AfterClass executes after the entire cucumber test run if you are using a Junit Runner. If you insert your feature tag into the runner, Junit will run all the scenarios in that feature and when it ends it execute the AfterClass. The irony is that If you try to run from your feature file directly Junit will not run..... The only way to run this is to run it via your Junit Runner via the tags... You can even extend this junit runner class with abstract TestNG cucumber Test Class and also include TestNG @BeforeClass and @AfterClass  this will run before and after the tags when you specify the class or suite in your testng.xml file.  


`@RunWith(Cucumber.class)
	@CucumberOptions (
			monochrome = true,
			features = "src/main/resources/features/Newtest", 
			plugin = {"pretty", "html:target/cucumber-html-report"},
			tags = "@editNewtest"
			)

	public class CukesRunner extends AbstractTestNGCucumberTests {
		
		@org.junit.BeforeClass 
		public static void junitsetup() {			
	        	loadandlogin();		
			}
		@org.junit.AfterClass
		public static void junitteardown(){
				browser.closeWebBrowser();
		}
			
		@org.testng.annotations.BeforeClass
		public static void testNGsetup() {	
        	loadandlogin();	
		}
		
		@@org.testng.annotations.AfterClass
		public static void testNGteardown(){
				browser.closeWebBrowser();
		}
		}
		`

**Comment by andresd24** [comment]
Hi @OlufemiAdeOlusile.   The @org.junit.AfterClass just doesn't want to run for me.  We solved this by extending the class to an abstract test class that adds a addShutdownHook to the Runtime and generates the reports....

TestClass.java:

`@RunWith(Cucumber.class)
@CucumberOptions(plugin = {"pretty","html:target/cucumber"})


public class IridiumDemoTest extends AbstractTest {

	  
}
`
AbstractTest.java:

`public abstract class AbstractTest {

	@BeforeClass
	public static void setUp(){
        Runtime.getRuntime().addShutdownHook(new Thread() {
            @Override
            public void run() {
            	try {
				//generate report;
				} catch (Exception e) {
					e.printStackTrace();
				}
            }
        });
    }
}`

Thanks for sharing TestNG since we'll look into using that in the future.   Hopefully we can get it to work the way you mention.

**Comment by OlufemiAdeOlusile** [comment]
@andresd24  your welcome, I don't think the problem is with the Junit Afterclass.. Rather look at your implementation again and find out if the afterclass gets the same instance after your tests run. Am sure this will be fine.. hehe..

**Comment by JWGmeligMeyling** [comment]
No, the workaround is not sufficient. This is because it only cleans up test state when the VM terminates, and not before continuing to execute other (JUnit) tests. Lets see if we can fix the problems with #678 .

**Comment by javapapo** [comment]
+1 for this issue

**Comment by nullexe** [comment]
+1 as well

This is super useful for testing an application, and ensuring that the app is closed out completely at the end of a test run.

**Comment by GoreA** [comment]
+1 fo rthis issue

**Comment by xoechts** [comment]
+1 for this issue

**Comment by FrankS77** [comment]
+1 for this issue

**Comment by jcannon-sovrn** [comment]
+1 for me here as well. I am going to try and make the problem go away with other means, but it would be really nice if Cucumber naturally supported this or made it otherwise more intuitive as to how to use it. I'm going to explain my issue in case that helps at all.
Currently using "Background" to load data into 3 different data stores in a Vagrant VM, then setting up Scenario Outlines for the purpose of exercising API endpoints with a variety of conditions, the API reading from the 3 data stores. It works, but as soon as I have the Background add multiple data sets (which are necessarily several thousand rows in some cases; not several thousand defined in the feature file, but by nature of the data being programmatically generated for Cassandra), the time to execute just blows up. The issue for me is that I don't necessarily want the same data set loaded for every feature file, which means in order to do Before/AfterClass, I'm going to have to get creative. Additionally, with some approaches I could lose the handy mapping of Cucumber->POJOs that align with my data structures, plus the readability of "what is the condition we're setting up before testing this set of endpoints." So I'm not able to just easily do one of the simple proposals here with Before/AfterClass.

Cucumber-Java has proven itself pretty popular at my company over the last year, but I'm basically forging the new territory by proving its usefulness, and this is the biggest application I've been tasked with sorting out. It's really nice to have something readable and clear like the Gherkin syntax, but this cumbersome lack of "set up the database scenario once before the scenario outline, not 20x for every example set of variables" is pretty awkward. I get that there are best practices and workarounds for most scenarios, but when you lose the clarity in the process...

**Comment by JWGmeligMeyling** [comment]
JBehave has a sane implementation of hooks. Unfortunately the JUnit runner is not entirely bug-free against newer versions which cause some issues in builds and IDE info. But it could be used as a reference implementation, or we could instead, address the issues there.

**Comment by kmelderis** [comment]
@AfterAll
public void afterAllScenarios () {
driver.quit();
}

Would be more readable (especially for those unlucky fresh apprentices who get thrown into 80k-of-lines maintenance test projects because management surely knows what they are doing); I haven't tried to implement it myself, but my vote goes for such a hook being present.

**Comment by richersoon** [comment]
+1

We need `@BeforeAll `or `@AfterAll `- because when we use Spring's `@Autowired` we are not able to use the  `@BeforeClass` that is static

**Comment by kmelderis** [comment]
To leave some value; Those that use JUnit as their runner (did not figure out how to do this with TestNG) you can use JUnit annotation to do something after "all your tests".

```java
@RunWith(Cucumber.class)
@CucumberOptions(
)
public class TestSuite {
    @AfterClass
    public static void teardown() {
        driver.quit();
    }
}
```

**Comment by Aaronstein** [comment]
+ 1 for the issue

**Comment by Easyman123** [comment]
+1 for this issue

**Comment by williamrussell-grindr** [comment]
I've been waiting for this forever. I can't believe it's still not supported. 

It's 2016 already.

**Comment by bugbare** [comment]
I've just come upon this issue, as I am tying to setup the load of some specific test data before a large series of scenarios for a specific feature, then I want to remove all data I have created for this feature at the end. 

At the moment, as a work around, I have created before/after steps that do the work above and are called by relevant tagging; 

e.g. '@load_test_data_for_feature_X' and '@remove_test_data_for_feature_X'. 

These are applied to empty scenarios at the start and end of the file, e.g.

@load_test_data_for_feature_X
scenario: PREP - load test data for feature X

@remove_test_data_for_feature_X
scenario: CLEANUP - remove test data for feature X

this seems to work, but as with all other posts above, would really like to see @Before_Feature and @After_Feature hooks, as these would be infinitely more flexible than my clumsy workaround.

**Comment by cosmocracy** [comment]
+1 because it's useful and reduces global code redundancy :-)   Plus code reviews will be nicer if we don't have to explain that though the frameworks supports Before it doesn't support BeforeAll.

**Comment by nullexe** [comment]
Additionally, when building a Cucumber instance that is triggered by code reviews, for example, being able to validate that all required systems have been correctly implemented and set up prior to execution is a godsend, and on the flip side, ensuring that same infrastructure has been correctly dismantled can prevent numerous issues on the build engineering side of a truly autonomous testing framework.

**Comment by andresd24** [comment]
totally. 100% agree that's the proactive and only way to do it.... it's
been done for a while but only by visionary engineers.

**Comment by jromero** [comment]
I'm grateful that @aslakhellesoy reverted the prior implementation because it would be a huge mistake not to support DI. My whole use case for needing these hooks is to use a class marked as singleton from the Injector for reporting purposes.

Unfortunately, I'm not yet fully versed in the whole concept of worlds and the internal implementation of the runner but hoping someone solves this in the near future.

+1 from me.

**Comment by linjin-harvey** [comment]
+1

**Comment by jizel** [comment]
+1

Still hoping...

**Comment by zhtaoang** [comment]
with multiple runners and multithreaded execution yet without an "afterAll" to close the driver thread pool is inconvenient, to say the least

**Comment by LoveleenSaini** [comment]
+1

**Comment by satheyogesh** [comment]
+1

**Comment by hammalas** [comment]
+1

**Comment by michaelHicks** [comment]
+1

**Comment by jasminprajapati** [comment]
+1

**Comment by depressiveRobot** [comment]
+1

**Comment by rnavarropiris** [comment]
++

**Comment by BogdanVovk** [comment]
+1

**Comment by jaesyoon** [comment]
+1

**Comment by roneygomes** [comment]
+1

**Comment by mdsina** [comment]
+1

**Comment by marinat** [comment]
+1

**Comment by kosteman** [comment]
+1

**Comment by houres** [comment]
+1

**Comment by marinat** [comment]
Is anybody of developers can comment this issue?

**Comment by vikramvi** [comment]
I've followed this thread with 
compile 'info.cukes:cucumber-java:1.2.5'
compile 'info.cukes:cucumber-junit:1.2.5'

wrt comments

1. https://github.com/cucumber/cucumber-jvm/issues/515#issuecomment-22057619

I found execution happens like BeforeClass -> Before -> After -> AfterClass

2. https://github.com/cucumber/cucumber-jvm/issues/515#issuecomment-22060842

But I found @ ClassRule does **not** get executed.
I've 2 methods in @ ClassRule
@Override
 protected void starting(Description desc)

@Override
protected void finished(Description desc)

cc: @wjpowell

**Comment by rvowles** [comment]
I have found the easiest way to do  this is to create a new cucumber "Backend" and then use buildWorld and disposeWorld methods to hook into.

**Comment by marinat** [comment]
@rvowles Could you share example code, please?

**Comment by mpkorstanje** [comment]
Another example solution using JUnit.

```
@RunWith(Cucumber.class)
public class Some02IT {

    @ClassRule
    public static TestRule classRule = (base, description) -> new Statement() {
        @Override
        public void evaluate() throws Throwable {
            try{
                setupExpensiveFixtures();
                base.evaluate();    
            } finally {
                tearDownExpensiveFixtures();
            }
        }
    };
}

```

**Comment by rvowles** [comment]
@marinat just create a class in the cucumber.runtime package (in your project) that implements the class "Backend" (from the same package). We implement our initialization in loadGlue and static cleanup (disposing of temporary AWS queues) in disposeWorld.

**Comment by k1ru** [comment]
+1

**Comment by cfreitas-ig** [comment]
+1
Still waiting.

**Comment by shnako** [comment]
+1

**Comment by aslakhellesoy** [comment]
I realise this has been open for a long while. Thanks everyone for the +1 comments.

One way to increase the likelihood of this happening is to give financial support to the volunteers who maintain Cucumber: https://opencollective.com/cucumber

**Comment by rvowles** [comment]
I thought I posted an answer to this? We used it to load all of our configuration, initialize log4j properly, and close down temporary SQS queues. Our test lead documents our e2e/cucumber stuff here: http://connect.cd/2017/10/introduction-to-microservice-test-automation-with-cucumber-jvm-and-openapi/

class **must** be in cumber.runtime package as that is all it scans. 

`package cucumber.runtime;

import cucumber.runtime.io.ResourceLoader;
import cucumber.runtime.snippets.FunctionNameGenerator;
import gherkin.formatter.model.Step;

import java.util.List;

public class BeforeAfterHooks implements Backend {
	private Glue glue;
	private List<String> gluePaths;

	public BeforeAfterHooks(ResourceLoader resourceLoader) {
	}

	@Override
	public void loadGlue(Glue glue, List<String> gluePaths) {
		this.glue = glue;
		this.gluePaths = gluePaths;
		
		// do your before hooks here
	}

	@Override
	public void setUnreportedStepExecutor(UnreportedStepExecutor executor) {

	}

	@Override
	public void buildWorld() {

	}

	@Override
	public void disposeWorld() {
		// do your after hooks here
	}

	@Override
	public String getSnippet(Step step, FunctionNameGenerator functionNameGenerator) {
		return null;
	}
}
`

**Comment by nsidhaye** [comment]
I wanted to clenup some data after all scenarios from feature file. I believe `@afterAll ("@tag")` will be ultimate solution.

I tried with `@After("@tag")` but it gets executed after every scenario from feature file.

**Comment by haroon-sheikh** [comment]
@mpkorstanje Do you think we can progress https://github.com/cucumber/cucumber-jvm/pull/678 this now? This was reverted as it was failing on DI.

**Comment by mpkorstanje** [comment]
I am not seeing a working pull request. And I have also not seen any fundamental discussion about how this is supposed to work with java8 and DI.

**Comment by pkokoshnikov** [comment]
+1 
it's really need for our testing. We use @BeforeClass @AfterClass TestNG to prepare data on cloud service for group of test. It's very expensive to create 'service project' for each scenario, thats why we want to use only one 'service project' for group of scenarios.

```
@Before
public void beforeAll() {
  if(!dunit) {
            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                if (project != null && Configuration.getInstance().isCloseProject()) {
                    try {
                        project.close();
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                }
            }));
            project = DnaNexusProject.create(getProjectName(), ImmutableMap.of(DATA_SOURCE_PROJECT, PermissionLevel.VIEW));
            dunit = true;
        }
``` 
This approach doesn't work, because addShutdownHook is called after first scenario.

**Comment by ctang3** [comment]
+1

**Comment by Silverrav** [comment]
+1

**Comment by Shurov** [comment]
+1

**Comment by 599139** [comment]
+1
In addition to@BeforeAll, can we have @BeforeFeature as well, something similar to Specflow? These hooks really makes framework maintenance easy.

Its really surprising to see this issue open since 2013!!

**Comment by pgajeshwar-coupa** [comment]
+1
These are very basic hooks and must have in cucumber.

**Comment by vikulin** [comment]
+1

**Comment by dhanainme** [comment]
+1

**Comment by symonk** [comment]
what is the best solution for runOnce on start / finish when running features in parallel on forked JVM processes?

**Comment by scormaq** [comment]
Hi there,
I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:

1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
```java
package com.foo.bar;

import cucumber.api.event.EventHandler;
import cucumber.api.event.EventPublisher;
import cucumber.api.event.TestRunFinished;
import cucumber.api.event.TestRunStarted;
import cucumber.api.formatter.Formatter;

public class Initialization implements Formatter {

    private EventHandler<TestRunStarted> setup = event -> {
        // setup static method(-s) here  
    };

    private EventHandler<TestRunFinished> teardown = event -> {
        // tear down static method(-s) here  
    };

    @Override
    public void setEventPublisher(EventPublisher publisher) {
        publisher.registerHandlerFor(TestRunStarted.class, setup);
        publisher.registerHandlerFor(TestRunFinished.class, teardown);
    }
}
```
2. Apply this plugin to Cucumber main class, e.g. like in next example:
```java
@RunWith(Cucumber.class)
@CucumberOptions(
        features = "src/test/resources/features/",
        glue = "com.foo.definitions",
        plugin = {"com.foo.bar.Initialization"})
public class CucumberRunner {
}
```
or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.

P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).

**Comment by fspl001147** [comment]
In Cucumber after last scenario the report is generated. I want to run few blocks of code after generation of report 
(say : I want to compress the folder which contains report and mail it.
**My question is as report is generated in the last of the program after last execution of last scenario and after class annotation. So where should I have to call that function which compress the report containing folder.)
Please suggest**

**Comment by michaelstanfa** [comment]
@fspl001147 did you find a solution to your issue? Also interested in similar functionality.

**Comment by ned29** [comment]
Hi, I am also interested the solution, we have a lot of steps classes where we use an embedded mongo, so will be very open connection once and close once

**Comment by Abba-** [comment]
While I would like to see this with a simple tag, @rvowles solution has worked perfectly for me, just needed to be updated (see below).  I needed to set System properties before running.  I threw them in and it works for me.

This makes it work for me running a normal IDEA feature test (Right click -> run feature).  I can't speak for @RunWith(cucumber.class).

Edited version as of July 2018 (Cucumber 2.4.0, Cucumber-Spring 2.3.0):

```
package cucumber.runtime; //cannot change.  can be under /test/java

import cucumber.runtime.io.ResourceLoader;
import cucumber.runtime.snippets.FunctionNameGenerator;
import gherkin.pickles.PickleStep;

import java.util.List;

public class WhateverYouWant implements Backend {
	private Glue glue;
	private List<String> gluePaths;

	public BeforeAfterHooks(ResourceLoader resourceLoader) {	}

	@Override
	public void loadGlue(Glue glue, List<String> gluePaths) {
		this.glue = glue;
		this.gluePaths = gluePaths;
		
		// do your before hooks here
	}

       @Override
	public void disposeWorld() {
		// do your after hooks here
	}

	@Override
	public void setUnreportedStepExecutor(UnreportedStepExecutor executor) {}

	@Override
	public void buildWorld() {}

	

	@Override
	public String getSnippet(PickleStep pickleStep, String s,, FunctionNameGeneratorfunctionNameGenerator) {
		return null;
	}
}
```

**Comment by fnrcum** [comment]
> 
> 
> Hi there,
> I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:
> 
>     1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
> 
> 
> ```java
> package com.foo.bar;
> 
> import cucumber.api.event.EventHandler;
> import cucumber.api.event.EventPublisher;
> import cucumber.api.event.TestRunFinished;
> import cucumber.api.event.TestRunStarted;
> import cucumber.api.formatter.Formatter;
> 
> public class Initialization implements Formatter {
> 
>     private EventHandler<TestRunStarted> setup = event -> {
>         // setup static method(-s) here  
>     };
> 
>     private EventHandler<TestRunFinished> teardown = event -> {
>         // tear down static method(-s) here  
>     };
> 
>     @Override
>     public void setEventPublisher(EventPublisher publisher) {
>         publisher.registerHandlerFor(TestRunStarted.class, setup);
>         publisher.registerHandlerFor(TestRunFinished.class, teardown);
>     }
> }
> ```
> 
>     1. Apply this plugin to Cucumber main class, e.g. like in next example:
> 
> 
> ```java
> @RunWith(Cucumber.class)
> @CucumberOptions(
>         features = "src/test/resources/features/",
>         glue = "com.foo.definitions",
>         plugin = {"com.foo.bar.Initialization"})
> public class CucumberRunner {
> }
> ```
> 
> or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.
> 
> P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).

I've tried this idea and it seems very interesting but my biggest question is, why exactly does the handler only run at the end of the suite?

**Comment by scormaq** [comment]
> I've tried this idea and it seems very interesting but my biggest question is, why exactly does the handler only run at the end of the suite?

I didn't get your question. What do you mean by "why exactly does the handler only run at the end of the suite"?
In my example there are 2 handlers - one for `TestRunStarted` event and one for `TestRunFinished` event. They are registered and ingested into Cucumber lifecycle via plugin mechanism. Because code in handlers executed on test run start/finish, we can emulate `before all` and `after all` behavior.

P.S. Also you'd better specify version of Cucumber dependencies that you use. My example was for 2.x version, in 4.x version some significant changes happened. I did not used it properly yet, but can assume that plugin now should implement [ConcurrentEventListener](https://github.com/cucumber/cucumber-jvm/blob/master/core/src/main/java/cucumber/api/event/ConcurrentEventListener.java), otherwise you can face unexpected behavior (see issue #1515 for more details).

**Comment by pushprajsingh05** [comment]
> Hi there,
> I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:
> 
> 1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
> 
> ```java
> package com.foo.bar;
> 
> import cucumber.api.event.EventHandler;
> import cucumber.api.event.EventPublisher;
> import cucumber.api.event.TestRunFinished;
> import cucumber.api.event.TestRunStarted;
> import cucumber.api.formatter.Formatter;
> 
> public class Initialization implements Formatter {
> 
>     private EventHandler<TestRunStarted> setup = event -> {
>         // setup static method(-s) here  
>     };
> 
>     private EventHandler<TestRunFinished> teardown = event -> {
>         // tear down static method(-s) here  
>     };
> 
>     @Override
>     public void setEventPublisher(EventPublisher publisher) {
>         publisher.registerHandlerFor(TestRunStarted.class, setup);
>         publisher.registerHandlerFor(TestRunFinished.class, teardown);
>     }
> }
> ```
> 
> 1. Apply this plugin to Cucumber main class, e.g. like in next example:
> 
> ```java
> @RunWith(Cucumber.class)
> @CucumberOptions(
>         features = "src/test/resources/features/",
>         glue = "com.foo.definitions",
>         plugin = {"com.foo.bar.Initialization"})
> public class CucumberRunner {
> }
> ```
> 
> or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.
> 
> P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).

It worked. Tested with tags, without tags, using CLI, without CLI (eclipse) ...

**Comment by alenmeister** [comment]
> Hi there,
> I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:
> 
> 1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
> 
> ```java
> package com.foo.bar;
> 
> import cucumber.api.event.EventHandler;
> import cucumber.api.event.EventPublisher;
> import cucumber.api.event.TestRunFinished;
> import cucumber.api.event.TestRunStarted;
> import cucumber.api.formatter.Formatter;
> 
> public class Initialization implements Formatter {
> 
>     private EventHandler<TestRunStarted> setup = event -> {
>         // setup static method(-s) here  
>     };
> 
>     private EventHandler<TestRunFinished> teardown = event -> {
>         // tear down static method(-s) here  
>     };
> 
>     @Override
>     public void setEventPublisher(EventPublisher publisher) {
>         publisher.registerHandlerFor(TestRunStarted.class, setup);
>         publisher.registerHandlerFor(TestRunFinished.class, teardown);
>     }
> }
> ```
> 
> 1. Apply this plugin to Cucumber main class, e.g. like in next example:
> 
> ```java
> @RunWith(Cucumber.class)
> @CucumberOptions(
>         features = "src/test/resources/features/",
>         glue = "com.foo.definitions",
>         plugin = {"com.foo.bar.Initialization"})
> public class CucumberRunner {
> }
> ```
> 
> or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.
> 
> P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).

How would you combine this with a common TestContext class that functions as World using DI (picocontainer). Currently we inject TestContext as an argument to our global Hooks constructor, but that would not work in Initialization because it only supports a selective list of constructor args.

**Comment by sirdir** [comment]
@pushprajsingh05 witch version of cucumber did you use?

**Comment by CreeTar** [comment]
since 8 years, people asking for this issue, now already a new better and more advanced framework is on the market. Cucumber is dead, long live Gauge!

**Comment by pushprajsingh05** [comment]
> @pushprajsingh05 witch version of cucumber did you use?

$ cucumber --version
*** WARNING: You must use ANSICON 1.31 or higher (https://github.com/adoxa/ansicon/) to get coloured output on Windows
3.1.2

**Comment by pushprajsingh05** [comment]
> Hi there,
> I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:
> 
> 1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
> 
> ```java
> package com.foo.bar;
> 
> import cucumber.api.event.EventHandler;
> import cucumber.api.event.EventPublisher;
> import cucumber.api.event.TestRunFinished;
> import cucumber.api.event.TestRunStarted;
> import cucumber.api.formatter.Formatter;
> 
> public class Initialization implements Formatter {
> 
>     private EventHandler<TestRunStarted> setup = event -> {
>         // setup static method(-s) here  
>     };
> 
>     private EventHandler<TestRunFinished> teardown = event -> {
>         // tear down static method(-s) here  
>     };
> 
>     @Override
>     public void setEventPublisher(EventPublisher publisher) {
>         publisher.registerHandlerFor(TestRunStarted.class, setup);
>         publisher.registerHandlerFor(TestRunFinished.class, teardown);
>     }
> }
> ```
> 
> 1. Apply this plugin to Cucumber main class, e.g. like in next example:
> 
> ```java
> @RunWith(Cucumber.class)
> @CucumberOptions(
>         features = "src/test/resources/features/",
>         glue = "com.foo.definitions",
>         plugin = {"com.foo.bar.Initialization"})
> public class CucumberRunner {
> }
> ```
> 
> or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.
> 
> P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).

Hi there, 

It was working for me and I was very happy, but one I got requirement to use parallel execution. I implemented it using below tutorial https://johnfergusonsmart.com/running-parallel-tests-serenity-bdd-cucumber/ and now it's not working.

Please advise.

**Comment by savita20** [comment]
> Hi there,
> I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:
> 
> 1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
> 
> ```java
> package com.foo.bar;
> 
> import cucumber.api.event.EventHandler;
> import cucumber.api.event.EventPublisher;
> import cucumber.api.event.TestRunFinished;
> import cucumber.api.event.TestRunStarted;
> import cucumber.api.formatter.Formatter;
> 
> public class Initialization implements Formatter {
> 
>     private EventHandler<TestRunStarted> setup = event -> {
>         // setup static method(-s) here  
>     };
> 
>     private EventHandler<TestRunFinished> teardown = event -> {
>         // tear down static method(-s) here  
>     };
> 
>     @Override
>     public void setEventPublisher(EventPublisher publisher) {
>         publisher.registerHandlerFor(TestRunStarted.class, setup);
>         publisher.registerHandlerFor(TestRunFinished.class, teardown);
>     }
> }
> ```
> 
> 1. Apply this plugin to Cucumber main class, e.g. like in next example:
> 
> ```java
> @RunWith(Cucumber.class)
> @CucumberOptions(
>         features = "src/test/resources/features/",
>         glue = "com.foo.definitions",
>         plugin = {"com.foo.bar.Initialization"})
> public class CucumberRunner {
> }
> ```
> 
> or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.
> 
> P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).

This solution worked for me as well.

Can you suggest me if I want to use this AfterClass annotation then how I can able to use it?

**Comment by hemanthsridhar** [comment]
https://medium.com/@hemanthsridhar92/global-hooks-in-cucumber-jvm-afc1be13e487
Please follow this tutorial

**Comment by hemanth-dunzo** [comment]
> what is the best solution for runOnce on start / finish when running features in parallel on forked JVM processes?
Using this way
https://medium.com/@hemanthsridhar92/global-hooks-in-cucumber-jvm-afc1be13e487

**Comment by noel-yap** [comment]
I'm not seeing how `@AfterAll` can be easily implemented by the user. At first I was thinking the `boolean` in the OP would have to become an `AtomicInteger` which is incremented each time an `@Before`-annotated method is called and decremented each time an `@After`-annotated method is called, but even then there are race conditions. For example, ScenarioA calls `@Before` then `@After` then ScenarioB calls `@Before` then `@After`. In this situation, both the meat of `@Before` and `@After` will be called twice.

Rather than advocating for Cucumber to implement `@BeforeAll` and `@AfterAll`, I'm thinking a better approach would be to reuse the features already existing in JUnit. This could even include replacing the implementations of `@Before` and `@After` to use JUnit's implementations. This might be even easier to do with JUnit 5.

**Comment by RayEltaib** [comment]
Can you please provide one for @AfterAll 
I need it to close the driver after all the testing is done I don't want to keep closing the driver every time after each scenario it wastes a lot of time.
Right now I achieve that via having a feature specifically for that but that creates a problem because my @After method has to deal with the UI.

**Comment by haroon-sheikh** [comment]
@RayEltaib Have you tried this yet https://terribletester.com/beforeall-and-afterall-workaround-for-cucumber-java?

**Comment by fstarred** [comment]
> > Hi there,
> > I came to next solution, that does not require hooks or `@BeforeClass`/etc JUnit annotations, maybe it will be useful for someone. It is valid for Java8 + `io.cucumber` model and its reworked plugin approach:
> > ```
> > 1. Create new plugin class by implementing `cucumber.api.formatter.Formatter` interface:
> > ```
> > 
> > 
> > ```java
> > package com.foo.bar;
> > 
> > import cucumber.api.event.EventHandler;
> > import cucumber.api.event.EventPublisher;
> > import cucumber.api.event.TestRunFinished;
> > import cucumber.api.event.TestRunStarted;
> > import cucumber.api.formatter.Formatter;
> > 
> > public class Initialization implements Formatter {
> > 
> >     private EventHandler<TestRunStarted> setup = event -> {
> >         // setup static method(-s) here  
> >     };
> > 
> >     private EventHandler<TestRunFinished> teardown = event -> {
> >         // tear down static method(-s) here  
> >     };
> > 
> >     @Override
> >     public void setEventPublisher(EventPublisher publisher) {
> >         publisher.registerHandlerFor(TestRunStarted.class, setup);
> >         publisher.registerHandlerFor(TestRunFinished.class, teardown);
> >     }
> > }
> > ```
> > 
> > 
> > ```
> > 1. Apply this plugin to Cucumber main class, e.g. like in next example:
> > ```
> > 
> > 
> > ```java
> > @RunWith(Cucumber.class)
> > @CucumberOptions(
> >         features = "src/test/resources/features/",
> >         glue = "com.foo.definitions",
> >         plugin = {"com.foo.bar.Initialization"})
> > public class CucumberRunner {
> > }
> > ```
> > 
> > 
> > or via command line option like `-Dcucumber.options="--plugin com.foo.bar.Initialization"` etc.
> > P.S. In this way we are able to implement not only BeforeAll/AfterAll behavior, but also interceptors for very different Cucumber events (before/after features, scenarios, steps), full list [here](https://github.com/cucumber/cucumber-jvm/tree/master/core/src/main/java/cucumber/api/event).
> 
> I've tried this idea and it seems very interesting but my biggest question is, why exactly does the handler only run at the end of the suite?

Nice job!
For the latest cucumber core, use **ConcurrentEventListener** instead of **Formatter** class

**Comment by ThanhTra1** [comment]
> Junit BeforeClass and AfterClass annotations work fine if you are using a JUnit runner. No issues with the static state.. Works fine with Page object Model for me.

Could you help me ? What should i do about it? I am stuck it a week and i dont have any ideal for it

**Comment by gbirchmeier** [comment]
@haroon-sheikh The whole `cucumber.api.event` package is gone after version 4.8.  Do you have an update?
